src/lib/InstantSearch.js

```diff
@@ -389,75 +312,69 @@ See: https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-o
     // later point.
     this.started = false;
 
-    this.derivedHelper.detach();
-    this.derivedHelper = null;
-
     // The helper needs to be reset to perform the next search from a fresh state.
     // If not reset, it would use the state stored before calling `dispose()`.
-    this.helper.removeAllListeners();
+    this.removeAllListeners();
+    this.mainHelper.removeAllListeners();
+    this.mainHelper = null;
     this.helper = null;
   }
 
+  scheduleSearch = defer(() => {
+    this.mainHelper.search();
+  });
+
+  scheduleRender = defer(() => {
+    if (!this.mainHelper.hasPendingRequests()) {
+      clearTimeout(this._searchStalledTimer);
+      this._searchStalledTimer = null;
+      this._isSearchStalled = false;
+    }
+
+    this.mainIndex.render({
+      instantSearchInstance: this,
+    });
+
+    this.emit('render');
+  });
+
+  scheduleStalledRender() {
+    if (!this._isSearchStalled && !this._searchStalledTimer) {
+      this._searchStalledTimer = setTimeout(() => {
+        this._isSearchStalled = true;
+        this.scheduleRender();
+      }, this._stalledSearchDelay);
+    }
+  }
+
   createURL(params) {
     if (!this._createURL) {
       throw new Error(
-        'The `start` method needs to be called before `createURL`.'
+        withUsage('The `start` method needs to be called before `createURL`.')
       );
     }
-    return this._createURL(this.helper.state.setQueryParameters(params));
-  }
 
-  _init(helper, state) {
-    this.widgets.forEach(widget => {
-      if (widget.init) {
-        widget.init({
-          state,
-          helper,
-          templatesConfig: this.templatesConfig,
-          createURL: this._createAbsoluteURL,
-          onHistoryChange: this._onHistoryChange,
-          instantSearchInstance: this,
-        });
-      }
-    });
+    return this._createURL(
+      this.mainIndex.getHelper().state.setQueryParameters(params)
+    );
   }
 
-  _render(helper, results, state) {
-    if (!this.helper.hasPendingRequests()) {
-      clearTimeout(this._searchStalledTimer);
-      this._searchStalledTimer = null;
-      this._isSearchStalled = false;
+  refresh() {
+    if (!this.mainHelper) {
+      throw new Error(
+        withUsage('The `start` method needs to be called before `refresh`.')
+      );
     }
 
-    this.widgets.forEach(widget => {
-      if (!widget.render) {
-        return;
-      }
-      widget.render({
-        templatesConfig: this.templatesConfig,
-        results,
-        state,
-        helper,
-        createURL: this._createAbsoluteURL,
-        instantSearchInstance: this,
-        searchMetadata: {
-          isSearchStalled: this._isSearchStalled,
-        },
-      });
-    });
-
-    /**
-     * Render is triggered when the rendering of the widgets has been completed
-     * after a search.
-     * @event InstantSearch#render
-     */
-    this.emit('render');
+    this.mainHelper.clearCache().search();
```

Comment:
This clears all the cache since there's only one search client, so we should be fine :)